daq/gateway.py

```diff
@@ -33,97 +30,89 @@ def __init__(self, runner, name, port_set, network):
         self.dhcp_monitor = None
         self.fake_target = None
         self.host = None
-        self.host_intf = None
         self.dummy = None
         self.tmpdir = None
         self.targets = {}
-        self.test_ports = {}
-        self.ready = {}
+        self.test_ports = set()
+        self.ready = set()
         self.activated = False
         self.result_linger = False
-        self._scan_monitor = None
+        self._dhcp_servers = {}
 
     def initialize(self):
         """Initialize the gateway host"""
         try:
             self._initialize()
         except Exception as e:
-            LOGGER.error('Gateway initialization failed, terminating: %s', str(e))
+            LOGGER.error(
+                'Gateway initialization failed, terminating: %s', str(e))
             self.terminate()
             raise
 
     def _initialize(self):
         host_name = 'gw%02d' % self.port_set
         host_port = self._switch_port(self.GATEWAY_OFFSET)
-        LOGGER.info('Initializing gateway %s as %s/%d', self.name, host_name, host_port)
-        self.tmpdir = self._setup_tmpdir(host_name)
-        cls = docker_host.make_docker_host('daqf/networking', prefix='daq', network='bridge')
-        # Work around an instability in the faucet/clib/docker library, b/152520627.
-        if getattr(cls, 'pullImage'):
-            setattr(cls, 'pullImage', lambda x: True)
-        host = self.runner.add_host(host_name, port=host_port, cls=cls, tmpdir=self.tmpdir)
-        host.activate()
-        self.host = host
-        self._change_lease_time(self.runner.config.get('initial_dhcp_lease_time'))
-        LOGGER.info("Added networking host %s on port %d at %s", host_name, host_port, host.IP())
+        LOGGER.info('Initializing gateway %s as %s/%d',
+                    self.name, host_name, host_port)
+        self.tmpdir = self._setup_tmpdir('inst', host_name)
+        self.host = self._start_dhcp_server(host_port).host
```

Comment:
As per previous discussion, the top-level primary construct here should be the gateway (or "networking" container if you prefer). For the simple case, DHCP should be a sub-component of that.